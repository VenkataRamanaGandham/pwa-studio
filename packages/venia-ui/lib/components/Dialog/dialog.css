.root {
    /* The Dialog takes up the entire screen and sits on top of the mask. */
    position: fixed;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
    background-color: transparent;
    z-index: 4;

    /* Set pointer-events to none to allow the mask to be clicked. */
    pointer-events: none;

    /*
     *  The Dialog is a grid container with its container
     *  horizontally and vertically centered.
     */
    display: grid;
    align-items: center;
    justify-items: center;
    align-content: center;

    /* The Dialog is hidden by default. */
    opacity: 0;
    visibility: hidden;

    /* The Dialog animates to being closed. */
    transition-duration: 192ms;
    transition-timing-function: var(--venia-anim-out);
    transition-property: opacity, visibility;
}

.root_open {
    composes: root;

    /* The Dialog animates to being open. */
    opacity: 1;
    visibility: visible;
    transition-duration: 224ms;
    transition-timing-function: var(--venia-anim-in);
}

.mask {
    /* The mask takes up and masks the entire screen. */
    position: fixed;
    top: 0;
    left: 0;
    height: 100vh;
    width: 100vw;
    background-color: rgb(var(--venia-grey-darker));
    opacity: 0.5;

    /* The Mask sits on top of all other content. */
    z-index: 3;

    /* The Mask is hidden by default. */
    opacity: 0;
    visibility: hidden;

    /* The Mask animates to being closed. */
    transition-duration: 192ms;
    transition-timing-function: var(--venia-anim-out);
    transition-property: opacity, visibility;
}

.mask_open {
    composes: mask;

    /* The Mask animates to being open. */
    opacity: 0.5;
    visibility: visible;
    transition-duration: 224ms;
    transition-timing-function: var(--venia-anim-in);
}

/* The container is the only grid item in the Dialog grid. */
.container {
    background-color: rgb(var(--venia-background-color));
    border-radius: 5px;
    box-shadow: 1px 1px 5px rgb(var(--venia-grey-darker));

    /* It sets maximum sizes so its body can handle overflow. */
    height: 100%;
    max-height: 90vh;
    /* Minimum keeps a 16:9 aspect ratio and is 40rem x 22.5rem.  */
    min-height: 360px;
    width: 640px;

    /* Nothing is allowed to overflow container itself. */
    overflow: hidden;

    /* Container is itself a grid container for its children. */
    display: grid;
    grid-template-rows: auto 1fr;

    /* Container can be the target of pointer events. */
    pointer-events: auto;
}

/*
 *  Dialog Header styles.
 */

.header {
    border-bottom: 1px solid rgb(var(--venia-border));
    height: 3.5rem;
    padding: 0 0.75rem 0 1rem;

    /* The Header is itself a grid container for its children. */
    display: grid;
    grid-auto-flow: column;
    grid-template-columns: 1fr auto;
    column-gap: 1rem;
    align-items: center;
}

.headerText {
    color: rgb(var(--venia-text-alt));
    text-transform: capitalize;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    line-height: 1.25rem;
}

.headerButton {
    /* Horizontally align the close button to the right. */
    justify-self: right;
}

/*
 *  Dialog Body styles.
 */

.body {
    padding: 1rem;
    overflow: auto;

    /* The Body is itself a grid container for its children. */
    display: grid;
    grid-template-rows: 1fr auto;
}

.buttons {
    padding: 2.5rem;

    display: grid;
    /* Force the buttons to be the same size (their content could differ). */
    grid-template-columns: 1fr 1fr;
    column-gap: 2rem;

    justify-self: center;
}

.button {
    overflow: hidden;
}

/*
 * Mobile-specific styles.
 *
 * Instead of being a full-page modal,
 * display within the confines of the target drawer.
 */

@media (max-width: 960px) {
    .root {
        left: auto;
        right: 0;
        max-width: 360px;

        /* The Dialog starts off-screen on the right. */
        transform: translate3d(100%, 0, 0);
        /* Animate the Dialog's position in mobile too. */
        transition-property: opacity, transform, visibility;
    }

    .root_open {
        composes: root;

        /* The Dialog animates (slides) onto the screen. */
        transform: translate3d(0, 0, 0);
    }

    .container {
        max-height: 100%;
    }

    .contents {
        /* The contents should be no wider than the body. */
        max-width: calc(360px - 2rem);
    }
}
