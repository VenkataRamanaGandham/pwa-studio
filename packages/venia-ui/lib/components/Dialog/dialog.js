import React from 'react';
import { bool, func, shape, string } from 'prop-types';
import { Form } from 'informed';
import { X as CloseIcon } from 'react-feather';

import { mergeClasses } from '@magento/venia-ui/lib/classify';

import Icon from '../Icon';
import { Modal } from '../Modal';
import defaultClasses from './dialog.css';

/**
 * The Dialog component shows its children content in a dialog,
 * encoding the look-and-feel and behavior in one place for consistency across the app.
 * 
 * @typedef Dialog
 * @kind functional component
 * 
 * @param {Object}  props
 * @param {Object}  props.classes - A set of class overrides to apply to elements.
 * @param {String}  props.cancelText - The text to display on the Dialog cancel button.
 * @param {String}  props.confirmText - The text to display on the Dialog confirm button.
 * @param {Func}    props.onCancel - A function to call when the user cancels the Dialog.
 * @param {Func}    props.onConfirm - A function to call when the user confirms the Dialog.
 * @param {Boolean} props.isModal - Determines behavior of clicking outside the content area. False cancels Dialog.
 * @param {Boolean} props.isOpen - Whether the Dialog is currently showing.
 * @param {String}  props.title - The title of the Dialog.
 */
const Dialog = props => {
    const {
        cancelText,
        children,
        confirmText,
        onCancel,
        onConfirm,
        isModal,
        isOpen,
        title,
    } = props;

    const classes = mergeClasses(defaultClasses, props.classes);
    const rootClass = isOpen ? classes.root_open : classes.root;

    return (
        <Modal>
            <aside className={rootClass}>
                <Form className={classes.contents} onSubmit={onConfirm}>
                    <div className={classes.header}>
                        <span className={classes.headerText}>
                            {title}
                        </span>
                        <button
                            className={classes.headerButton}
                            onClick={onCancel}
                            type="reset"
                        >
                            <Icon src={CloseIcon} />
                        </button>
                    </div>
                    <div className={classes.body}>
                        {children}
                        <div className={classes.buttons}>
                            <button
                                className={classes.cancelButton}
                                onClick={onCancel}
                                type="reset"
                            >
                                {cancelText}
                            </button>
                            <button
                                className={classes.confirmButton}
                                type="submit"
                            >
                                {confirmText}
                            </button>
                        </div>
                    </div>
                </Form>
            </aside>
        </Modal>
    );
};

export default Dialog;

Dialog.propTypes = {
    classes: shape({
        body: string,
        cancelButton: string,
        confirmButton: string,
        contents: string,
        header: string,
        headerText: string,
        headerButton: string,
        root: string,
        root_open: string,
    }),
    cancelText: string,
    confirmText: string,
    onCancel: func,
    onConfirm: func,
    isModal: bool,
    isOpen: bool,
    title: string
};

Dialog.defaultProps = {
    cancelText: 'Cancel',
    confirmText: 'Confirm',
    isModal: false
};
