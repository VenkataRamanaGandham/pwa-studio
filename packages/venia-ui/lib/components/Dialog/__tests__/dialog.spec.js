import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import Dialog from '../dialog';

jest.mock('../../../classify');
jest.mock('../../Modal', () => ({
    Modal: props => <modal-mock>{props.children}</modal-mock>
}));

const props = {
    isModal: true,
    isOpen: true,
    shouldDisableButtons: false,
    title: 'Unit Test Dialog Title'
};

test('renders a basic Dialog', () => {
    // Act.
    const wrapper = createTestInstance(<Dialog {...props} />);

    // Assert.
    expect(wrapper.toJSON()).toMatchSnapshot();
});

test('renders a Dialog with disabled buttons', () => {
    // Arrange.
    const myProps = {
        ...props,
        shouldDisableButtons: true
    };

    // Act.
    const wrapper = createTestInstance(<Dialog {...myProps} />);

    // Assert.
    expect(wrapper.toJSON()).toMatchSnapshot();
});

test('supports modifying title and button texts', () => {
    // Arrange.
    const myProps = {
        ...props,
        cancelText: 'Unit Test Cancel',
        confirmText: 'Unit Test Confirm',
        title: 'Unit Test Title'
    };

    // Act.
    const wrapper = createTestInstance(<Dialog {...myProps} />);

    // Assert.
    expect(wrapper.toJSON()).toMatchSnapshot();
});
