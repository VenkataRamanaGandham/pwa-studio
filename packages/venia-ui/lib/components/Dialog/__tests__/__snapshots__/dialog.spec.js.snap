// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render a close X button in modal mode 1`] = `
<modal-mock>
  <button
    disabled={true}
    type="button"
  />
  <aside>
    <form
      onKeyDown={[Function]}
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div>
        <span>
          Unit Test Dialog Title
        </span>
      </div>
      <div>
        <div />
        <div>
          <button
            disabled={false}
            type="reset"
          >
            Cancel
          </button>
          <button
            disabled={false}
            type="submit"
          >
            Confirm
          </button>
        </div>
      </div>
    </form>
  </aside>
</modal-mock>
`;

exports[`renders a Dialog with disabled buttons 1`] = `
<modal-mock>
  <button
    disabled={true}
    type="button"
  />
  <aside>
    <form
      onKeyDown={[Function]}
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div>
        <span>
          Unit Test Dialog Title
        </span>
        <button
          disabled={true}
          type="reset"
        >
          <span
            className="root"
          >
            <svg
              className="icon"
              fill="none"
              height="24"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="18"
                x2="6"
                y1="6"
                y2="18"
              />
              <line
                x1="6"
                x2="18"
                y1="6"
                y2="18"
              />
            </svg>
          </span>
        </button>
      </div>
      <div>
        <div />
        <div>
          <button
            disabled={true}
            type="reset"
          >
            Cancel
          </button>
          <button
            disabled={true}
            type="submit"
          >
            Confirm
          </button>
        </div>
      </div>
    </form>
  </aside>
</modal-mock>
`;

exports[`renders a basic Dialog 1`] = `
<modal-mock>
  <button
    disabled={false}
    type="button"
  />
  <aside>
    <form
      onKeyDown={[Function]}
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div>
        <span>
          Unit Test Dialog Title
        </span>
        <button
          disabled={false}
          type="reset"
        >
          <span
            className="root"
          >
            <svg
              className="icon"
              fill="none"
              height="24"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="18"
                x2="6"
                y1="6"
                y2="18"
              />
              <line
                x1="6"
                x2="18"
                y1="6"
                y2="18"
              />
            </svg>
          </span>
        </button>
      </div>
      <div>
        <div />
        <div>
          <button
            disabled={false}
            type="reset"
          >
            Cancel
          </button>
          <button
            disabled={false}
            type="submit"
          >
            Confirm
          </button>
        </div>
      </div>
    </form>
  </aside>
</modal-mock>
`;

exports[`supports modifying title and button texts 1`] = `
<modal-mock>
  <button
    disabled={false}
    type="button"
  />
  <aside>
    <form
      onKeyDown={[Function]}
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div>
        <span>
          Unit Test Title
        </span>
        <button
          disabled={false}
          type="reset"
        >
          <span
            className="root"
          >
            <svg
              className="icon"
              fill="none"
              height="24"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="18"
                x2="6"
                y1="6"
                y2="18"
              />
              <line
                x1="6"
                x2="18"
                y1="6"
                y2="18"
              />
            </svg>
          </span>
        </button>
      </div>
      <div>
        <div />
        <div>
          <button
            disabled={false}
            type="reset"
          >
            Unit Test Cancel
          </button>
          <button
            disabled={false}
            type="submit"
          >
            Unit Test Confirm
          </button>
        </div>
      </div>
    </form>
  </aside>
</modal-mock>
`;
